### PR TITLE
feat: add pester testcases

### LINKED ISSUE #161: Add `logging output to a file` feature in pester testcases
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Add logging output to a file feature in pester testcases

### Use Case(s)

Right now pester testcases development for testing PasswordPolicyManagment code is going on. Once initial thing is phased out, it is better to log all messages in to a file, which helps for debugging.

### Potential Configuration

writing to a log file instead of just dumping the output on the console.

### References

_No response_

### LINKED ISSUE #162: Add pester testcases for automated testing of the code
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Today we have manual testing process to test the VMware.CloudFoundation.PasswordManagement code, but this has to be put in CD/CI pipeline to automate the testing process for repeatable testing before pushing to public GitHub to test on the quality of the fix and to minimize regression defects. This also ensures, time consumed for less in testing the code.

### Use Case(s)

Suppose if manual testing takes about x time, automation of the same takes x/2. Also it is more reliable and you can test even negative testcases easily.

### Potential Configuration

writing pester testcases to automate the testing process
In the current code structure, there will be /tests folder, under that pester testcases will be available for testing
Also it requires input json file to hold all the credential information

### References

na

### LINKED ISSUE #163: Enhance pester testcases with tagging for running group of testcases at a time
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

To run a bunch of testcases together in pester, tagging is the only mechanism. So add tags to testcases so as to group them and run the group

### Use Case(s)

Suppose I want to run only positive testcases in my testbed, in which case, use 
`Invoke-Pester -Path <ScriptPath> -Tag 'positive'`
This runs all testcases tagged as 'positive'

### Potential Configuration

You have to add -Tag parameter at `It` blocks or `Describe` blocks or at `Context` blocks.

### References

na

### PR DESCRIPTION
**Pull request by sowjuec**
### Summary

 Pester testcases are added to test the password management code and integrate to CD/CI pipeline

### Type

- [ ] Bugfix
- [X] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### Test and Documentation

- [X] Tests have been completed.
- [ ] Documentation has been added or updated.

### Issue References

Closes #162 
Closes #161 
Closes #163

### CODE REVIEW

File: tests/logs/PPM.Tests.ps1-02-27-2024_07_40_28.log
Comment (tenthirtyam) [inline tests/logs/PPM.Tests.ps1-02-27-2024_07_40_28.log:1]: 1. I'd recommend adding a `.gitkeep` file in `./tests/logs` so that it keeps the "empty" directory but default but does not include any of our logs.
2. Update .gitignore to exclude `/tests/logs/`.

```console
# MacOS
## Ignore desktop services files.
**/.DS_Store

# Pester
## Ignore logs directory.
**/tests/logs/**

# MkDocs
## Ignore site directory.
**/.site/**

File: tests/InputData.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		],
+		"Manager Nodes" : [
+				"sfo-w01-nsx01a.sfo.rainpole.io",
+				"sfo-w01-nsx01b.sfo.rainpole.io",
+				"sfo-w01-nsx01c.sfo.rainpole.io"
+		],
+		"Cluster"	:	["sfo-w01-cl01"]
+	},
+	"sfo-m01": {
+		"Esxi Server" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:38]:
```suggestion
		"ESXi Hosts" : [
```

Could we update this?

File: tests/InputData.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+						"sfo-m01"
+					],
+	"sfo-w01": {
+		"Esxi Server" : [
+							"sfo01-w01-esx01.sfo.rainpole.io",
+							"sfo01-w01-esx02.sfo.rainpole.io",
+							"sfo01-w01-esx03.sfo.rainpole.io",
+							"sfo01-w01-esx04.sfo.rainpole.io"
+						],
+		"Edge Nodes" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:23]:
```suggestion
		"NSX Edge Nodes" : [
```

Could we update this?

File: tests/InputData.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+							"sfo-w01-en01",
+							"sfo-w01-en02"
+		],
+		"Manager Node" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:27]:
```suggestion
		"NSX Manager" : [
```

```suggestion
		"NSX Edge Nodes" : [
```

Could we update this?

File: tests/InputData.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+			"sfo-w01-nsx01.sfo.rainpole.io"
+		],
+		"Manager Nodes" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:30]:
```suggestion
		"NSX Manager Nodes" : [
```

Could we update this?

File: tests/InputData.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	"Password" 	: "VMw@re1!",
+	"SsoAdmin"	: "admin@local",
+	"SsoPass"	: "VMw@re1!VMw@re1!",
+	"RootPass"	: "VMw@re1!",
+	"Domain"	: 	[
+						"sfo-w01",
+						"sfo-m01"
+					],
+	"sfo-w01": {
+		"Esxi Server" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:17]:
```suggestion
		"ESXi Hosts" : [
```

Could we update this?

File: tests/InputData.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		"Cluster"	:	["sfo-w01-cl01"]
+	},
+	"sfo-m01": {
+		"Esxi Server" : [
+			"sfo01-m01-esx01.sfo.rainpole.io",
+			"sfo01-m01-esx02.sfo.rainpole.io",
+			"sfo01-m01-esx03.sfo.rainpole.io",
+			"sfo01-m01-esx04.sfo.rainpole.io"
+		],
+		"Edge Nodes" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:44]:
```suggestion
		"NSX Edge Nodes" : [
```

Could we update this?

File: tests/InputData.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+					"sfo-m01-en01",
+					"sfo-m01-en02"
+		],
+		"Manager Node" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:48]:
```suggestion
		"NSX Manager" : [
```

Could we update this?

File: tests/InputData.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+			"sfo-m01-nsx01.sfo.rainpole.io"
+		],
+		"Manager Nodes" : [
```
Comment (tenthirtyam) [inline tests/InputData.json:51]:
```suggestion
		"NSX Manager Nodes" : [
```

Could we update this?

File: tests/InputData.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+			"sfo-m01-nsx01a.sfo.rainpole.io",
+			"sfo-m01-nsx01b.sfo.rainpole.io",
+			"sfo-m01-nsx01c.sfo.rainpole.io"
+		],
+		"Cluster"	:	["sfo-m01-cl01"]
```
Comment (tenthirtyam) [inline tests/InputData.json:56]:
```suggestion
		"Clusters"	:	[
			"sfo-m01-cl01"
		]
```

Could we update this since it can take >1.

File: tests/InputData.json [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		],
+		"Manager Node" : [
+			"sfo-w01-nsx01.sfo.rainpole.io"
+		],
+		"Manager Nodes" : [
+				"sfo-w01-nsx01a.sfo.rainpole.io",
+				"sfo-w01-nsx01b.sfo.rainpole.io",
+				"sfo-w01-nsx01c.sfo.rainpole.io"
+		],
+		"Cluster"	:	["sfo-w01-cl01"]
```
Comment (tenthirtyam) [inline tests/InputData.json:35]:
```suggestion
		"Clusters"	:	[
			"sfo-w01-cl01"
		]
```

Could we update this since it can take >1.

File: tests/logs/PPM.Tests.ps1-02-27-2024_07_40_28.log
Comment (tenthirtyam) [inline tests/logs/PPM.Tests.ps1-02-27-2024_07_40_28.log:1]: Delete this log file. 😄

File: tests/PPM.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,1650 @@
...
+
+                    # Attempt to update the ESXi host password expiration settings.
+                    $updateResult = Update-EsxiPasswordExpiration -server $server -user $user -pass $pass -domain  $domain -cluster $cluster -maxDays $invalidMaxDays
+
+                    # Output the update result.
+                    Write-LogToFile -message "Update Result: $updateResult"
+
+                    # If the function did not throw an error, fail the test.
+                    $null | Should -Be $updateResult
+
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
```

File: tests/PPM.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,1650 @@
...
+                    
+                    # Get the updated Max Days.
+                    $outMaxDays = $updatedExpirationSettings[0].'Max Days'
+
+                    # Output the updated Max Days.
+                    Write-LogToFile -message "Updated Max Days: $outMaxDays"
+
+                    # Assert that the updated Max Days is equal to the Decremented Max Days.
+                    $outMaxDays | Should -Be $maxDays
+
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
```

File: tests/PPM.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,1650 @@
...
+                    # Attempt to update the ESXi host password expiration settings.
+                    $updateResult = Update-SsoPasswordExpiration -server $server -user $user -pass $pass -domain $domain -maxDays $invalidMaxDays
+
+                    # Output the update result.
+                    Write-LogToFile -message "Update Result: $updateResult"
+
+                    # If the function did not throw an error, fail the test. If setting is already present, it will be skipped and result will be null
+                    $null | Should -Be  $updateResult
+
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
```

File: tests/PPM.Tests.ps1 [hunk 14]
```diff
@@ -0,0 +1,1650 @@
...
+            BeforeEach {
+                # Request the current virtual center password expiration settings
+                $currentExpirationSettings = Request-VcenterPasswordExpiration -server $server -user $user -pass $pass -domain $domain
+            
+                # Increment the values by 1.
+                $minDays = [int]$currentExpirationSettings[0].'Min Days' + 1
+                $warnDays = [int]$currentExpirationSettings[0].'Warning Days' + 1
+                $maxDays = [int]$currentExpirationSettings[0].'Max Days' + 1
+
+            }
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:249]:
```suggestion
            }
```

File: tests/PPM.Tests.ps1 [hunk 15]
```diff
@@ -0,0 +1,1650 @@
...
+                } catch {
+                    Write-LogToFile -Type ERROR -message "An error occurred: $_"
+                    $false | Should -Be $true
+                } finally {
+                    Write-LogToFile -message "End of virtual center Password Expiration Positive Testcase"
+                }
+            }
+
+            # Expect a failure.
+            #Accepting negative value, So gave bigger value
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
            # Accepting negative value, so gave bigger value.
```

File: tests/PPM.Tests.ps1 [hunk 16]
```diff
@@ -0,0 +1,1650 @@
...
+
+                    # If an error was thrown, fail the test.
+                    $false | Should -Be $true
+                } finally {
+                    Write-LogToFile -message "End of local user Password Expiration negative Testcase"
+                }
+            }
+        }
+        
+        Describe 'NSX edge Password Expiration'  -Tag "NSXEdgePasswordExpiration" {
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
        Describe 'NSX Edge Password Expiration'  -Tag "NSXEdgePasswordExpiration" {
```

File: tests/PPM.Tests.ps1 [hunk 17]
```diff
@@ -0,0 +1,1650 @@
...
+                    
+                } catch {
+                    Write-LogToFile -Type ERROR -message "An error occurred: $_"
+                    $false | Should -Be $true
+                } finally {
+                    Write-LogToFile -message "End of NSX edge Password Expiration positive Testcase" 
+                }
+            }
+
+            #Accepted range of value is <1-9999>
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
            # Accepted range of value is <1-9999>
```

File: tests/PPM.Tests.ps1 [hunk 18]
```diff
@@ -0,0 +1,1650 @@
...
+                    Write-LogToFile -message "Updated Max Days: $outMaxDays"
+
+                    # Assert that the updated Max Days is equal to the Decremented Max Days.
+                    $outMaxDays | Should -Be $maxDays
+                    
+                } catch {
+                    Write-LogToFile -Type ERROR -message "An error occurred: $_"
+                    $false | Should -Be $true
+                } finally {
+                    Write-LogToFile -message "End of NSX edge Password Expiration positive Testcase"
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1]:
```suggestion
                    Write-LogToFile -message "End of NSX Edge Password Expiration positive Testcase" 
```

File: tests/PPM.Tests.ps1
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:1]: 1. `split-path` >> `Split-Path`
2. `join-path` >> `Join-Path`
3. `Write-Host  -NoNewline` >> `Write-Host -NoNewline`
4. `try` >> `Try`
5. `catch` >> `Catch`
6. `finally` >> `Finally`

File: tests/PPM.Tests.ps1 [hunk 19]
```diff
@@ -0,0 +1,1638 @@
...
+        }
+
+        Start-SetupLogFile
+        $InputData = Get-Content -Raw 'InputData.json' | ConvertFrom-Json
+        $server = $InputData.'SDDCManager'
+        $user = $InputData.'User'
+        $pass = $InputData.'Password'
+        $domain = $InputData.'Domain'[0]
+        $esxiServer = $InputData.$domain.'Esxi Server'[0]
+        $cluster = $InputData.$domain.'Cluster'[0]
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:93]:
```suggestion
        $cluster = $InputData.$domain.'Clusters'[0]
```

File: tests/PPM.Tests.ps1 [hunk 20]
```diff
@@ -0,0 +1,1638 @@
...
+            }
+        }
+
+        Start-SetupLogFile
+        $InputData = Get-Content -Raw 'InputData.json' | ConvertFrom-Json
+        $server = $InputData.'SDDCManager'
+        $user = $InputData.'User'
+        $pass = $InputData.'Password'
+        $domain = $InputData.'Domain'[0]
+        $esxiServer = $InputData.$domain.'Esxi Server'[0]
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:92]:
```suggestion
        $esxiServer = $InputData.$domain.'ESXi Hosts'[0]
```

File: tests/PPM.Tests.ps1 [hunk 21]
```diff
@@ -0,0 +1,1638 @@
...
+                $index = $output.'Index'
+            }
+            if (-Not $flag) {
+                Write-LogToFile -Type ERROR -message "$server or $user is not matching in the $output"
+            }
+        }
+
+        Start-SetupLogFile
+        $InputData = Get-Content -Raw 'InputData.json' | ConvertFrom-Json
+        $server = $InputData.'SDDCManager'
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:88]:
```suggestion
        $server = $InputData.'SDDC Manager'
```

File: tests/PPM.Tests.ps1 [hunk 22]
```diff
@@ -0,0 +1,1638 @@
...
+        $pass = $InputData.'Password'
+        $domain = $InputData.'Domain'[0]
+        $esxiServer = $InputData.$domain.'Esxi Server'[0]
+        $cluster = $InputData.$domain.'Cluster'[0]
+        $vmName = $InputData.'SDDCManagerVMName'
+        $rootPass = $InputData.'RootPass'
+        $rootUser = $InputData.'RootUser'
+        $guestUser = $InputData.'GuestUser'
+        $localUser = $InputData.'LocalUser'
+        $nsxManagerNode = $InputData.$domain.'Manager Nodes'[0]
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:99]:
```suggestion
        $nsxManagerNode = $InputData.$domain.'NSX Manager Nodes'[0]
```

File: tests/PPM.Tests.ps1 [hunk 23]
```diff
@@ -0,0 +1,1638 @@
...
+        $nsxEdgeNode = $InputData.$domain.'Edge Nodes'[0]
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:100]:
```suggestion
        $nsxEdgeNode = $InputData.$domain.'NSX Edge Nodes'[0]
```

File: tests/InputData.json [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+	"SDDCManagerVMName" : "sfo-vcf01",
+	"User"	:	"administrator@vsphere.local",
+	"RootUser"	: "root",
+	"GuestUser" : "root",
+	"LocalUser" : "vcf",
+	"Password" 	: "VMw@re1!",
+	"SsoAdmin"	: "admin@local",
+	"SsoPass"	: "VMw@re1!VMw@re1!",
+	"RootPass"	: "VMw@re1!",
+	"Domain"	: 	[
```
Comment (tenthirtyam) [inline tests/InputData.json:12]:
```suggestion
	"Domains"	: 	[
```

File: tests/InputData.json [hunk 25]
```diff
@@ -0,0 +1,58 @@
+{		
+	"SDDCManager" : "sfo-vcf01.sfo.rainpole.io",
```
Comment (tenthirtyam) [inline tests/InputData.json:2]:
```suggestion
	"SDDC Manager" : "sfo-vcf01.sfo.rainpole.io",
```

File: tests/InputData.json [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+	"SDDCManagerVMName" : "sfo-vcf01",
```
Comment (tenthirtyam) [inline tests/InputData.json:3]:
```suggestion
	"SDDC Manager VM Name" : "sfo-vcf01",
```

File: tests/InputData.json [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+	"User"	:	"administrator@vsphere.local",
+	"RootUser"	: "root",
+	"GuestUser" : "root",
+	"LocalUser" : "vcf",
+	"Password" 	: "VMw@re1!",
+	"SsoAdmin"	: "admin@local",
```
Comment (tenthirtyam) [inline tests/InputData.json:9]:
```suggestion
	"SsoAdmin"	: "administrator@local",
```

File: tests/PPM.Tests.ps1 [hunk 28]
```diff
@@ -0,0 +1,1638 @@
...
+
+        Start-SetupLogFile
+        $InputData = Get-Content -Raw 'InputData.json' | ConvertFrom-Json
+        $server = $InputData.'SDDCManager'
+        $user = $InputData.'User'
+        $pass = $InputData.'Password'
+        $domain = $InputData.'Domain'[0]
+        $esxiServer = $InputData.$domain.'Esxi Server'[0]
+        $cluster = $InputData.$domain.'Cluster'[0]
+        $vmName = $InputData.'SDDCManagerVMName'
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:94]:
```suggestion
        $vmName = $InputData.'SDDC Manager VM Name'
```

File: tests/PPM.Tests.ps1 [hunk 29]
```diff
@@ -0,0 +1,1638 @@
...
+                Write-LogToFile -Type ERROR -message "$server or $user is not matching in the $output"
+            }
+        }
+
+        Start-SetupLogFile
+        $InputData = Get-Content -Raw 'InputData.json' | ConvertFrom-Json
+        $server = $InputData.'SDDCManager'
+        $user = $InputData.'User'
+        $pass = $InputData.'Password'
+        $domain = $InputData.'Domain'[0]
```
Comment (tenthirtyam) [inline tests/PPM.Tests.ps1:91]:
```suggestion
        $domain = $InputData.'Domains'[0]
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
I've performed some cleanup and left the changes to the JSON inputs for your review and comment.

File: tests/InputData.json (hunk 4 above)
Comment (sowjuec) [reply tests/InputData.json:30]: yes for sure

File: tests/InputData.json (hunk 5 above)
Comment (sowjuec) [reply tests/InputData.json:17]: sure

**Review by tenthirtyam — APPROVED** [review]
